### ISSUE TITLE
Conditional Kernel Version Check for BPFtrace Programs

### ISSUE THREAD

**Issue by benkilimnik** [iovisor/bpftrace#2729]
### Is your feature request related to a problem? Please describe.

As the Linux kernel evolves, certain kernel features, functions, or probes may change, be deprecated, or removed. This can lead to BPFtrace programs that rely on specific kernel features to break over time, particularly those relying on kprobes. For example, the `kprobe:tcp_drop` was removed in Linux kernel version 5.19, breaking the `tcp_drops` script (see #1895). 

Projects like [Pixie](https://github.com/pixie-io/pixie), which use BPFtrace for dynamic tracing, need the ability to deploy different BPFtrace programs based on the kernel version of the nodes in a cluster.

### Describe the solution you'd like

Introduce a conditional check in BPFtrace that allows programs to be deployed only if the target system's kernel version matches a specified range. This would enable users to write BPFtrace programs that are compatible with multiple kernel versions and ensure that only the appropriate version of the program is deployed on each node.

For example, a user could write two versions of a BPFtrace program, one for kernel versions up to 5.18 and another for kernel versions 5.19 and above. The BPFtrace runtime would then check the kernel version of the target system and deploy the appropriate version of the program.

The syntax for specifying the kernel version range could be similar to the following:

```bpftrace
@kernel(>=5.19)
{
  // BPFtrace program for kernel versions 5.19 and above
}

@kernel(<5.19)
{
  // BPFtrace program for kernel versions up to 5.18
}
```

Integrating a conditional kernel version check into BPFtrace would provide a more robust and flexible way to deploy BPFtrace programs in heterogeneous environments with varying kernel versions. 

### Describe alternative solutions or features you've considered

An alternative solution would be to handle the kernel version check externally, before deploying the BPFtrace program. However, this would require additional tooling and complexity for the user and would not be as seamless as having the kernel version check integrated into BPFtrace.

**Comment by danobi** [comment]
Agreed that feature probing is more robust. Distros do backports all the time without necessarily bumping the version code.

Perhaps something like weak symbols for probes might make sense. In that a failure to attach is silent. Not sure if we'd want to support multiple weak attachments to the same probe, though. Perhaps needs to be fleshed out.

**Comment by viktormalik** [comment]
> Perhaps something like weak symbols for probes might make sense. In that a failure to attach is silent. Not sure if we'd want to support multiple weak attachments to the same probe, though. Perhaps needs to be fleshed out.

Just for reference, we're doing something similar with wildcarded and multi-attachpoint probes now, although not silently. If there are multiple attachpoints for a probe and some of them don't exist, we show a warning and still attach to the rest. If there is a single attachpoint, we fail hard if it doesn't exist.

**Comment by leitao** [comment]
Another datapoint for discussion. Sometimes the semantics change, not only the hook point names. For example, my company has a script that checks for `RQF_STATS` macro on a block request completion. Something as:

```
rawtracepoint:block_rq_complete {
   if ($req_flags & _RQF_STATS`)
      do_work();

```

That flag changed in version 6.11 due to `5f89154e8e9e344 ("block: Use enum to define RQF_x bit indexes")`
and that flag value is the same, so, I need to know exactly what kernel version I am running to use the correct flag value.|

**Comment by danobi** [comment]
@leitao it should be possible to access kernel version through header files. Not super clean, but it works:

```
#include <linux/version.h>

BEGIN {
  printf("%d.%d.%d\n", LINUX_VERSION_MAJOR, LINUX_VERSION_PATCHLEVEL, LINUX_VERSION_SUBLEVEL);
}
```

```
$ sudo bpftrace /tmp/bt.bt
Attaching 1 probe...
6.12.15
^C
```
